### PR TITLE
refactoring and bug fix

### PR DESCRIPTION
**Pull request by schweitzpgi**

### CODE REVIEW

File: flang/lib/Lower/IntrinsicCall.cpp [hunk 1]
```diff
@@ -1152,25 +1152,25 @@ mlir::Value IntrinsicLibrary::genIAnd(mlir::Type resultType,
 mlir::Value IntrinsicLibrary::genIchar(mlir::Type resultType,
                                        llvm::ArrayRef<mlir::Value> args) {
   // There can be an optional kind in second argument.
-  assert(args.size() >= 1);
+  assert(args.size() >= 1 && args.size() <= 2);
```
Comment (vdonaldson) [inline flang/lib/Lower/IntrinsicCall.cpp:1155]: Is there a nice way to do something like

assert(auto size = args.size(), size >=1 && size <= 2)

Comment (schweitzpgi) [reply flang/lib/Lower/IntrinsicCall.cpp:1155]: That might work. Not sure how portable it would be though.

Comment (sscalpone) [reply flang/lib/Lower/IntrinsicCall.cpp:1155]: Declare size() as a pure function :)

Comment (schweitzpgi) [reply flang/lib/Lower/IntrinsicCall.cpp:1155]: https://llvm.org/doxygen/ArrayRef_8h_source.html#l00156

It's already an inline const getter. A decent C++ compiler should be able to figure it out.

**Pull request merged**